File: java/log-levels/src/main/java/LogLevels.java
import java.util.regex.Pattern;

public class LogLevels {

  private static Pattern pattern = Pattern.compile("^\\[(?<level>[a-zA-Z_\\-0-9]+)\\]\\s*:\\s*(?<message>\\S.*)\\s*$");

  public static String message(String logLine) {
    var matches = pattern.matcher(logLine);

    return matches.find()
        ? matches.group("message").trim()
        : null;
  }

  public static String logLevel(String logLine) {
    var matches = pattern.matcher(logLine);

    return matches.find()
        ? matches.group("level").trim().toLowerCase()
        : null;
  }

  public static String reformat(String logLine) {
    var matches = pattern.matcher(logLine);

    return matches.find()
        ? String.format(
            "%s (%s)",
            matches.group("message").trim(),
            matches.group("level").trim().toLowerCase())
        : null;
  }
}

File: cpp/log-levels/log_levels.cpp
#include <string>

namespace log_line {
    std::string message(std::string log_message) {
       return log_message.substr(log_message.find(": ") + 2);
    }

    std::string log_level(std::string log_message) {
       return log_message.substr(1, log_message.find(": ") - 2);
    }

    std::string reformat(std::string log_message) {
       return message(log_message) + " (" + log_level(log_message) + ")";
    }
} // namespace log_line
